### PR TITLE
Fix regressions in cast from string to date and timestamp

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Partially addresses https://github.com/NVIDIA/spark-rapids/issues/3478

cuDF recently started supporting variable-length date/timestamp components in `is_timestamp` and when parsing strings as timestamps. This caused regressions in the plugin.

This PR fixes the regressions and has the following changes:

- Removes code that modified strings to replace single digit month and day with two digit versions
- Removes methods based on assumptions around fixed-lengh parsing in cuDF
- Updates regex patterns to accept one or two-digit month, day, hour, minute, second components
- Enables a test for casting string to timestamp since the only existing tests were ignored due to other documented issues

We also now support single digit hour, minute and second in timestamps. We still require 6 digits for fractional second.

### CODE REVIEW

**Comment by andygrove** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala [hunk 1]
```diff
@@ -960,30 +959,6 @@ class CastOpSuite extends GpuExpressionTestSuite {
     }
   }
 
-  test("CAST string to date - sanitize step") {
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala:963]: This test is no longer needed since we no longer have the sanitization step

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
The last build failure looked like a networking issue (could not fetch from github...)

**Pull request merged**